java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java

```diff
@@ -0,0 +1,204 @@
+      TypeSubstitution newSubstitution = inferTypeSubstitution(method, substitution, formalType.elementType, newArgType, variableArity, remainingArgTypes);
+      return mergeTypeSubstitutions(substitution, newSubstitution);
+    }
+    return substitution;
+  }
+
+  private static JavaType leastUpperBound(List<JavaType> remainingArgTypes) {
+    return (JavaType) Types.leastUpperBound(mapToBoxedSet(remainingArgTypes));
+  }
+
+  private static Set<Type> mapToBoxedSet(List<JavaType> types) {
+    return Sets.newHashSet(Iterables.transform(Sets.<Type>newHashSet(types), new Function<Type, Type>() {
+      @Override
+      public Type apply(Type type) {
+        if (type.isPrimitive()) {
+          return ((JavaType) type).primitiveWrapperType;
+        }
+        return type;
+      }
+    }));
+  }
+
+  private TypeSubstitution inferTypeSubstitutionInParameterizedType(MethodJavaSymbol method, TypeSubstitution substitution, ParametrizedTypeJavaType formalType, JavaType argType,
+    boolean variableArity, List<JavaType> remainingArgTypes) {
+    List<JavaType> formalTypeSubstitutedTypes = formalType.typeSubstitution.substitutedTypes();
+    TypeSubstitution result = substitution;
+    if (argType.isParameterized()) {
+      List<JavaType> argTypeSubstitutedTypes = ((ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
+      TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
+      result = mergeTypeSubstitutions(substitution, newSubstitution);
+    } else if (isRawTypeOfType(argType, formalType) || isNullType(argType)) {
+      List<JavaType> objectTypes = listOfTypes(symbols.objectType, formalTypeSubstitutedTypes.size());
+      TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, objectTypes);
+      result = mergeTypeSubstitutions(substitution, newSubstitution);
+    } else if (argType.isSubtypeOf(formalType.erasure()) && argType.isClass()) {
+      for (JavaType superType : ((ClassJavaType) argType).symbol.superTypes()) {
+        if (sameErasure(formalType, superType)) {
+          TypeSubstitution newSubstitution = inferTypeSubstitution(method, substitution, formalType, superType, variableArity, remainingArgTypes);
+          result = mergeTypeSubstitutions(substitution, newSubstitution);
+          break;
+        }
+      }
+    }
+    return result;
+  }
+
+  private static boolean isRawTypeOfType(JavaType rawType, JavaType type) {
+    return rawType == type.erasure();
+  }
+
+  private static List<JavaType> listOfTypes(JavaType type, int size) {
+    List<JavaType> result = new ArrayList<>(size);
+    for (int j = 0; j < size; j++) {
+      result.add(type);
+    }
+    return result;
+  }
+
+  private static boolean sameErasure(JavaType type1, JavaType type2) {
+    return type1.erasure() == type2.erasure();
+  }
+
+  private boolean isNullType(JavaType type) {
+    return type == symbols.nullType;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)